[PGO] Add `llvm.loop.estimated_trip_count` metadata

jdenny-ornl opened the pull request:
This patch implements the `llvm.loop.estimated_trip_count` metadata discussed in [[RFC] Fix Loop Transformations to Preserve Block Frequencies](https://discourse.llvm.org/t/rfc-fix-loop-transformations-to-preserve-block-frequencies/85785). As [suggested in the RFC comments](https://discourse.llvm.org/t/rfc-fix-loop-transformations-to-preserve-block-frequencies/85785/4), it adds the new metadata to all loops at the time of profile ingestion and estimates each trip count from the loop's `branch_weights` metadata.  As [suggested in the PR #128785 review](https://github.com/llvm/llvm-project/pull/128785#discussion_r2151091036), it does so via a new `PGOEstimateTripCountsPass` pass, which creates the new metadata for each loop but omits the value if it cannot estimate a trip count due to the loop's form.

An important observation not previously discussed is that `PGOEstimateTripCountsPass` *often* cannot estimate a loop's trip count, but later passes can sometimes transform the loop in a way that makes it possible.  Currently, such passes do not necessarily update the metadata, but eventually that should be fixed.  Until then, if the new metadata has no value, `llvm::getLoopEstimatedTripCount` disregards it and tries again to estimate the trip count from the loop's current `branch_weights` metadata.

jdenny-ornl committed:
[PGO] Add `llvm.loop.estimated_trip_count` metadata

This patch implements the `llvm.loop.estimated_trip_count` metadata
discussed in [[RFC] Fix Loop Transformations to Preserve Block
Frequencies](https://discourse.llvm.org/t/rfc-fix-loop-transformations-to-preserve-block-frequencies/85785).
As [suggested in the RFC
comments](https://discourse.llvm.org/t/rfc-fix-loop-transformations-to-preserve-block-frequencies/85785/4),
it adds the new metadata to all loops at the time of profile ingestion
and estimates each trip count from the loop's `branch_weights`
metadata.  As [suggested in the PR#128785
review](https://github.com/llvm/llvm-project/pull/128785#discussion_r2151091036),
it does so via a `PGOEstimateTripCountsPass` pass, which creates the
new metadata for the loop but omits the value if it cannot estimate a
trip count due to the loop's form.

An important observation not previously discussed is that
`PGOEstimateTripCountsPass` *often* cannot estimate a loop's trip
count but later passes can transform the loop in a way that makes it
possible.  Currently, such passes do not necessarily update the
metadata, but eventually that should be fixed.  Until then, if the new
metadata has no value, `llvm::getLoopEstimatedTripCount` disregards it
and tries again to estimate the trip count from the loop's
`branch_weights` metadata.

mtrofin reviewed (commented):
still reading through, and sorry for the delay on my end, I was on vacation. One high level question, should this maybe also touch the bitcode validator?

jdenny-ornl committed:
Merge branch 'main' into pgo-estimated-trip-count

jdenny-ornl committed:
Add PGOEstimateTripCounts in more cases

jdenny-ornl committed:
Add unused initialization

jdenny-ornl commented:
> still reading through, and sorry for the delay on my end, I was on vacation.

No worries.  I'm still catching up after a vacation as well.

> One high level question, should this maybe also touch the bitcode validator?

Are you suggesting something similar to your ProfileVerifierPass?

jdenny-ornl commented on llvm/lib/Transforms/Utils/LoopUtils.cpp:
```diff
@@ -834,33 +865,86 @@ static std::optional<unsigned> getEstimatedTripCount(BranchInst *ExitingBranch,
+
+  // Return the estimated trip count from metadata unless the metadata is
+  // missing or has no value.
+  bool Missing;
```
Done.

jdenny-ornl commented on llvm/test/Other/new-pm-defaults.ll:
```diff
@@ -131,31 +134,43 @@
 ; CHECK-O-NEXT: Running pass: RequireAnalysisPass<{{.*}}ProfileSummaryAnalysis
 ; CHECK-O-NEXT: Running analysis: InnerAnalysisManagerProxy
 ; CHECK-O-NEXT: Running analysis: LazyCallGraphAnalysis
+; CHECK-O-NEXT: Running analysis: TargetLibraryAnalysis
```
I must confess that I'm updating these pass-list tests by blindingly copying from the actual output.  I hope that some reviewer finds meaning in these updates, which are painful.

nikic replied:
The need to make substantial changes to these files was basically the indication that something is wrong :)

jdenny-ornl replied:
Thanks.  Converting it a function pass as you suggested indeed made the diff much smaller.

jdenny-ornl committed:
Simplify some test changes

jdenny-ornl committed:
Extend verify pass to cover new metadata

jdenny-ornl commented on llvm/lib/Transforms/Utils/LoopUtils.cpp:
```diff
@@ -201,34 +201,40 @@ void llvm::initializeLoopPassPass(PassRegistry &Registry) {
-          if (IntMD && IntMD->getSExtValue() == V)
-            // It is already in place. Do nothing.
-            return;
+          // If it is already in place, do nothing.
```
I think they meant for "it" to refer to the exact metadata, including any value.  How about the following?

```suggestion
          // If the metadata and any value are already as specified, do nothing.
```

jdenny-ornl commented on llvm/include/llvm/Transforms/Utils/LoopUtils.h:
```diff
@@ -316,28 +318,73 @@ LLVM_ABI TransformationMode hasDistributeTransformation(const Loop *L);
+/// Return either:
+/// - The value of \c llvm.loop.estimated_trip_count from the loop metadata of
+///   \p L, if that metadata is present and has a value.
+/// - Else, a new estimate of the trip count from the latch branch weights of
```
> should getLoopEstimatedTripCount just return what's in the metadata, and this recalculation be handled separately, with an update of the metadata?

Possibly, but not yet.  See the new LangRef.rst entry's "Missing Metadata and Values" section.

Let's converge on this issue before I address your other comments, many of which seem to stem from it.

jdenny-ornl replied:
> Ah, so the idea is to have the implementation here in this PR as a stop-gap, correct?

That sounds right for one side of it: some passes need to be adjusted to maintain this metadata properly.

But I am also thinking of any existing test cases and other use cases that don't strictly follow the approach we envision (i.e., add branch weight metadata to every loop, and then run pgo-estimate-trip-counts to add all estimated trip count metadata).  For example:
1. What if there are already branch weights and opt is called without pgo-estimate-trip-counts but with passes that call getLoopEstimatedTripCount or setLoopEstimatedTripCount?
2. When, e.g., debugging a broken pass, would running pgo-estimate-trip-counts be useful for fixing up just the loops that have missing estimated trip counts (either metadata or just values)?

For now, I'd like to keep things flexible, and LLVM_DEBUG can tell us when usage doesn't fit the ideal vision.

> So subsequent PRs would fix passes to write the metadata, and eventually `getLoopEstimatedTripCount` would just fetch metadata?

Possibly, depending on at least the following:
1. Who will identify and fix all such passes?  I'm not eager to keep putting effort into estimated trip counts right now.  First, some in the community think estimated trip counts are not worthwhile generally.  Second, I came here to work on block frequency issues, and the goal of this PR is to move estimated trip counts out of the way of that effort. 
2. As we work with the new metadata, will the greater flexibility of the current PR benefit use cases like those I mentioned above?

> 
> Maybe also add to the TODO in `getLoopEstimatedTripCount` that "...and rely solely on metadata" or smth?

Sure, I can add a note there.

mtrofin replied:
OK, we can evolve this as its value becomes more apparent.

jdenny-ornl committed:
Fix test for some builds

Somehow, on some of my builds, `llvm::` prefixes are dropped from some
symbol names in the printed past list.

jdenny-ornl committed:
Merge branch 'main' into pgo-estimated-trip-count

jdenny-ornl committed:
Apply some small reviewer suggestions

jdenny-ornl committed:
Attempt to fix windows pre-commit CI

jdenny-ornl committed:
Merge branch 'main' into pgo-estimated-trip-count

jdenny-ornl commented:
Thanks for the review.

nikic commented on llvm/lib/Transforms/Instrumentation/PGOEstimateTripCounts.cpp:
```diff
@@ -0,0 +1,45 @@
+      MadeChange |= runOnLoop(L);
+  }
+  LLVM_DEBUG(dbgs() << DEBUG_TYPE << ": end\n");
+  return MadeChange ? PreservedAnalyses::none() : PreservedAnalyses::all();
```
This is invalidating analysis results for the entire module.

This should be a function pass with a function to module pass adaptor. And the function pass can preserve the CFG analysis group at least.

jdenny-ornl replied:
Thanks.  Applied in PR #152775.

nikic reviewed (commented):
This causes a substantial compile-time regression: https://llvm-compile-time-tracker.com/compare.php?from=3e579d93ab50952628a51bda05f3a39f6a5a631c&to=f7b65011de519b1bd987892475db61f99dde44ce&stat=instructions%3Au

The new pass looks somewhat problematic to me, from a pipeline perspective. Even if the whole-world invalidation is fixed, this is still going to compute DT and LI in a pretty random part of the optimization pipeline, where these will probably get invalidated before they can be reused.

I'm not super clear on what the benefits of this approach with an early inference pass are. I like this *conceptually*, but if in practice we can usually only infer the estimated trip counts after loop rotation, is the early pass really useful?

It's probably worth pointing out that just the presence of loop metadata impacts optimization behavior (in an attempt to preserve the metadata), so inserting dummy metadata for all loops will have secondary effects.

I'd probably suggest reverting this for now, we need to at least sort out the impact on the optimization pipeline.

mtrofin commented:
> Thanks for the explanation. I think I have it reverted properly.
> 
> @mtrofin What are your thoughts?

Could this be enabled behind a flag? IIRC there are a few changes needed here, besides this one; after which presumably we could try it out on some benchmarks.

jdenny-ornl commented:
> Less of an issue than the compile time regression, but this also introduced a layering violation between IR and TransformUtils. ~every file in lib/Transforms/Utils/ uses llvm/IR/*.h, and this commit added the reverse dependency as now llvm/lib/IR/Verifier.cpp includes llvm/Transforms/Utils/LoopUtils.h.

That just enables the verifier and utils to share `LLVMLoopEstimatedTripCount = "llvm.loop.estimated_trip_count"`.  What's the right way to do that?

> Could this be enabled behind a flag? IIRC there are a few changes needed here, besides this one; after which presumably we could try it out on some benchmarks.

If I understood correctly, the problems @nikic pointed out are all related to the new pass.  Even with this PR, `getLoopEstimatedTripCount` still estimates trip counts from branch weights when the metadata is missing, so my patches to fix BFI issues do not need the pass.  For my purposes then, yes, the pass can go behind a flag so that the pass can be investigated as an orthogonal issue.

> I'm not super clear on what the benefits of this approach with an early inference pass are. I like this _conceptually_, but if in practice we can usually only infer the estimated trip counts after loop rotation, is the early pass really useful?

@mtrofin Would you comment on this point?

> It's probably worth pointing out that just the presence of loop metadata impacts optimization behavior (in an attempt to preserve the metadata), so inserting dummy metadata for all loops will have secondary effects.

One possibility is for the pass to add the new metadata only to loops that have branch weights.  That way, you would only pay for it when you are doing PGO-like work.  The PR currently adds the metadata without a value in that case.

rupprecht commented:
> > Less of an issue than the compile time regression, but this also introduced a layering violation between IR and TransformUtils. ~every file in lib/Transforms/Utils/ uses llvm/IR/*.h, and this commit added the reverse dependency as now llvm/lib/IR/Verifier.cpp includes llvm/Transforms/Utils/LoopUtils.h.
> 
> That just enables the verifier and utils to share `LLVMLoopEstimatedTripCount = "llvm.loop.estimated_trip_count"`. What's the right way to do that?
> 

I really don't know what the best way to do this is, but here are a couple possibilities:

* Define `LLVMLoopEstimatedTripCount` in IR and have TransformUtils get it from there.
* Put `LLVMLoopEstimatedTripCount` in some lower level library (or create a new one) that both IR and TransformUtils depend on (e.g. Support)
* Just duplicate the string constant, possibly w/ a test to ensure they're the same value
* Pull Verifier.cpp out into a separate library (e.g. create IRVerifier), assuming the rest of IR doesn't depend on it.

btw, see also https://llvm.org/docs/CodingStandards.html#library-layering

jdenny-ornl commented:
> I really don't know what the best way to do this is, but here are a couple possibilities:
> 
>     * Define `LLVMLoopEstimatedTripCount` in IR and have TransformUtils get it from there.

Thanks.  I went with that in PR #152775.

jdenny-ornl commented on llvm/lib/Transforms/Instrumentation/PGOEstimateTripCounts.cpp:
```diff
@@ -0,0 +1,45 @@
+  return MadeChange;
+}
+
+PreservedAnalyses PGOEstimateTripCountsPass::run(Module &M,
```
In PR #152775, I went with the suggestion to [make it a function pass](https://github.com/llvm/llvm-project/pull/148758/files#r2246193446).
